File: python/day5.py
with open("day5.txt") as txt:
    input = txt.read().strip().split("\n\n")
    input[0] = input[0].split("\n")
    input[1] = input[1].split("\n")

ranges = [[int(j) for j in  i.split("-")] for i in input[0]]
fruits = [int(i) for i in input[1]]

def part1():
    count = 0

    for i in fruits:
        for j in ranges:
            if i >= j[0] and i <= j[1]:
                count += 1
                break

    return count

def part2():
    rnges = sorted(ranges)
    all_ranges = [rnges[0]]

    for i in range(1, len(rnges)):
        if rnges[i][0] <= all_ranges[-1][1]:
            all_ranges[-1] = [min(*all_ranges[-1], *rnges[i]), max(*all_ranges[-1], *rnges[i])]

        else:
            all_ranges.append(rnges[i])

    count = 0
    for i in all_ranges:
        count += i[1]-i[0]+1
    return count

print(part1())
print(part2())

File: ocaml/day5.ml
let input =
  String.split_on_char '\n'
    (In_channel.with_open_text "./day5.txt" In_channel.input_all)

let ranges =
  input
  |> List.filter (fun x -> String.contains x '-')
  |> List.map (fun x ->
      match String.split_on_char '-' x with
      | [ a; b ] -> (int_of_string a, int_of_string b)
      | _ -> failwith "Invalid input format")

let fruits =
  input
  |> List.filter (fun x ->
      List.fold_left
        (fun acc y -> acc || String.contains x y)
        false
        (List.init 10 (fun x -> char_of_int (x + int_of_char '0')))
      && not (String.contains x '-'))
  |> List.map int_of_string

let part1 fruits ranges =
  List.fold_left
    (fun acc x ->
      acc
      +
      if
        List.fold_left
          (fun acc (rng_start, rng_end) ->
            acc || (x >= rng_start && x <= rng_end))
          false ranges
      then 1
      else 0)
    0 fruits

let part2 ranges =
  (ranges
  |> List.sort (fun (a_start, a_end) (b_start, b_end) -> a_start - b_start)
  |> List.fold_left
       (fun acc (rng_start, rng_end) ->
         let last_start, last_end = List.hd acc in
         if rng_start > last_end then [ (rng_start, rng_end) ] @ acc
         else
           [
             ( List.fold_left min Int.max_int
                 [ rng_start; rng_end; last_start; last_end ],
               List.fold_left max 0 [ rng_start; rng_end; last_start; last_end ]
             );
           ]
           @ List.drop 1 acc)
       [ (0, 0) ]
  |> List.fold_left
       (fun acc (rng_start, rng_end) -> rng_end - rng_start + 1 + acc)
       0)
  - 1

let time f x =
  let t = Sys.time () in
  let fx = f x in
  Printf.printf "%d (%fs)\n%!" fx (Sys.time () -. t)

let time2 f x y =
  let t = Sys.time () in
  let fx = f x y in
  Printf.printf "%d (%fs)\n%!" fx (Sys.time () -. t)

let () = time2 part1 fruits ranges
let () = time part2 ranges
